### ISSUE TITLE
CLS loss off the charts while fine-tuning yolov8n-seg!

### ISSUE THREAD

**Issue by Gaurav-Shah05** [ultralytics/ultralytics#15769]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi there! I am experimenting with fine-tuning the yolov8n-seg model weights with different number representations like posits. 
The current training pipeline generates only FP32 weights, which gives options to quantize them further to FP16 and INT8. I tried these quantization schemes with openvino to make my model deployable on edge (running the test on the Intel Nezha board currently...) for real-time instance segmentation. The INT8 quantization scheme worked moderately well, but [posits](https://spectrum.ieee.org/floating-point-numbers-posits-processor) promises even higher inference speeds by lowering the weight precisions. 
Therefore, I am trying to use the pre-trained yolov8n-seg model and using the [qtorch+framework](https://dl.acm.org/doi/10.1007/978-3-031-09779-9_3) 
To train the fixed-posit weights. I have integrated qtorch in the yolo-training pipeline by updating the BaseTrainer class in the trainer.py file present in ultralytics/engine. I have attached the updated file here (the code added is present between #------- string, haven't removed any existing code).
[trainer.txt](https://github.com/user-attachments/files/16724160/trainer.txt)
The custom OptimLP optimizer definition:
'
class OptimLP(Optimizer):
    """
    A low-precision optimizer wrapper that handles weight, gradient, accumulator quantization.

    Args:
        - :attr: `optim`: underlying optimizer to use
        - :attr: `weight_quant`: a weight quantization function which takes a pytorch tensor and returns a tensor. If None, does not quantize weight.
        - :attr: `grad_quant`: a gradient quantization function which takes a pytorch tensor and returns a tensor. If None, does not quantize weight.
        - :attr: `grad_scaling`: float, scaling factor before apply gradient quantization.
        - :attr: `momentum_quant`: a momentum quantization function which takes a pytorch tensor and returns a tensor.
                                   If None, does not quantize weight.
        - :attr: `acc_quant`: a accumulator quantization function which takes
                              a pytorch tensor and returns a tensor. If not None, a
                              OptimLP object would create memory copies of model parameters that serve as
                              gradient accumulators. If None, does not use gradient accumulators.

    Example:
        >>> weight_q = quantizer(...) # define weight quantization
        >>> optimizer = SGD(model.parameters(), lr=0.1, momentum=0.9)
        >>> optimizer = OptimLP(optiimizer, weight_quant=weight_q)
    """

    def __init__(
        self,
        optim,
        weight_quant=None,
        grad_scaling=1.0,
        grad_quant=None,
        momentum_quant=None,
        acc_quant=None,
    ):
        assert isinstance(optim, SGD) or isinstance(optim, Adam)
        super(OptimLP, self).__init__(
            optim.param_groups, optim.defaults
        )  # place holder

        # python dictionary does not copy by default
        self.param_groups = optim.param_groups
        self.optim = optim

        assert grad_scaling > 0, "gradient scaling must be positive"
        self.grad_scaling = grad_scaling

        self.weight_quant = weight_quant
        self.grad_quant = grad_quant
        self.momentum_quant = momentum_quant
        self.acc_quant = acc_quant

        if isinstance(self.optim, SGD):
            self.momentum_keys = ["momentum_buffer"]
        elif isinstance(self.optim, Adam):
            # TODO: support amsgrad
            self.momentum_keys = ["exp_avg", "exp_avg_sq"]
        else:
            raise NotImplementedError("Only supporting Adam and SGD for now. ")

        if self.acc_quant != None:
            self.weight_acc = {}
            for group in self.param_groups:
                for p in group["params"]:
                    self.weight_acc[p] = p.detach().clone()

    def step(self, closure=None):
        """
        Performs one step of optimization with the underlying optimizer.
        Quantizes gradient and momentum before stepping. Quantizes gradient accumulator and weight after stepping.
        """
        # quantize gradient
        if not self.grad_quant is None:
            for group in self.param_groups:
                for p in group["params"]:
                    if p.grad is not None:  # Add this check
                        p.grad.data = self.grad_quant(p.grad.data * self.grad_scaling)

        # switch acc into weight before stepping
        if not self.acc_quant is None:
            for group in self.param_groups:
                for p in group["params"]:
                    p.data = self.weight_acc[p].data

        loss = self.optim.step()

        # switch weight into acc after stepping and quantize
        if not self.acc_quant is None:
            for group in self.param_groups:
                for p in group["params"]:
                    p.data = self.weight_acc[p].data = self.acc_quant(p.data).data

        # quantize weight from acc
        if not self.weight_quant is None:
            for group in self.param_groups:
                for p in group["params"]:
                    p.data = self.weight_quant(p.data).data

        # quantize momentum
        if not self.momentum_quant is None:
            for group in self.param_groups:
                if isinstance(self.optim, SGD) and group["momentum"] == 0:
                    continue
                for p in group["params"]:
                    if p.grad is not None:  # Add this check
                        param_state = self.optim.state[p]
                        for key in self.momentum_keys:
                            param_state[key] = self.momentum_quant(param_state[key])

        return loss

    def __repr__(self):
        return "LP Optimizer: {}".format(self.optim.__repr__())

    def __str__(self):
        return "LP Optimizer: {}".format(self.optim.__str__())`
        
  Coming back to the problem (reason why you probably opened this issue), when training the model using 
  `yolo mode=train model=yolov8n-seg.pt data=yolov7/data/coco.yaml imgsz=640 epochs=10 batch=16`
  in the first epoch, the cls_loss is off the charts:
 
![image](https://github.com/user-attachments/assets/8a74923f-1d6b-4010-a3d3-c4d61ad326fd)

What might the reason for that?

**Comment by Gaurav-Shah05** [comment]
@glenn-jocher I hope you can look into this issue and provide Insights. Please let me know if you need more details to help determine what went wrong. Thanks in advance!

**Comment by glenn-jocher** [comment]
@Gaurav-Shah05 thank you for sharing the details. It seems the high cls_loss might be due to the custom modifications. Please verify if the issue persists with the latest Ultralytics package and standard configurations. If the problem continues, consider checking your quantization implementation for potential issues.
